validator/src/journal/chain.rs

```diff
@@ -0,0 +1,683 @@
+            committed_transactions_count.inc_n(result.transaction_count);
+
+            let mut block_num_guage = COLLECTOR.gauge("ChainController.block_num", None, None);
+            block_num_guage.set_value(chain_head_block.block_num());
+
+            let chain_head = state.chain_head.clone().unwrap();
+            notify_on_chain_updated(
+                &mut state,
+                chain_head,
+                result.committed_batches,
+                result.uncommitted_batches,
+            );
+
+            state.chain_head.as_ref().map(|block| {
+                block.batches().iter().for_each(|batch| {
+                    if batch.trace {
+                        debug!(
+                            "TRACE: {}: ChainController.on_block_validated",
+                            batch.header_signature
+                        )
+                    }
+                })
+            });
+
+            let mut new_chain = result.new_chain;
+            new_chain.reverse();
+
+            for block in new_chain {
+                let receipts: Vec<TransactionReceipt> = block
+                    .execution_results
+                    .iter()
+                    .map(TransactionReceipt::from)
+                    .collect();
+                for observer in state.observers.iter_mut() {
+                    observer.chain_update(&block, &receipts.iter().collect::<Vec<_>>());
+                }
+            }
+            let total_committed_txns = match state.chain_reader.count_committed_transactions() {
+                Ok(count) => count,
+                Err(err) => {
+                    error!(
+                        "Unable to read total committed transactions count: {:?}",
+                        err
+                    );
+                    0
+                }
+            };
+
+            let mut committed_transactions_gauge =
+                COLLECTOR.gauge("ChainController.committed_transactions_gauge", None, None);
+            committed_transactions_gauge.set_value(total_committed_txns);
+        } else {
+            info!("Rejected new chain head: {}", new_block);
+        }
+    }
+
+    /// Light clone makes a copy of this controller, without access to the stop
+    /// handle.
+    pub fn light_clone(&self) -> Self {
+        ChainController {
+            state: self.state.clone(),
+            // This instance doesn't share the stop handle: it's not a
+            // publicly accessible instance
+            stop_handle: Arc::new(Mutex::new(None)),
+            block_queue_sender: self.block_queue_sender.clone(),
+            validation_result_sender: self.validation_result_sender.clone(),
+        }
+    }
+
+    fn submit_blocks_for_verification(
+        &self,
+        block_validator: &BV,
+        blocks: &[BlockWrapper],
+    ) -> Result<(), ChainControllerError> {
+        if self.validation_result_sender.is_some() {
```

Comment:
This function will silently do nothing if the sender is none, which may make debugging more difficult in the future. Can this be an expect() or a log message to help us out in the event it becomes none for some reason?

Reply:
Same with queue block

Reply:
Validation results should be an `expect`, since it can only be run after the controller has been started.  Queue block cannot, since it is exposed to python, and therefore control of the method is beyond the scope of this PR